### ISSUE TITLE
0.25 Release

### ISSUE THREAD

**Issue by bluejekyll** [hickory-dns/hickory-dns#2206]
I want to start tracking what we need for the next release. I'm thinking it might be a good idea to start releasing some alphas, as I think we're waiting on a few upstream things. Thoughts?

- [x] upgrade `ring`
- [x] upgrade `quinn`
... others

**Comment by djc** [comment]
I noticed that we now also depend on h3, so I started a PR for upgrading h3-quinn to Quinn 0.11 already.

**Comment by djc** [comment]
Draft PR for the TLS upates in #2217.

**Comment by AaronKutch** [comment]
the `smallvec` version and some other utility crates are quite old, could they be updated if it doesn't conflict with MSRV?

**Comment by djc** [comment]
@AaronKutch they should! If you want to help out with that, it would be greatly appreciated!

**Comment by AaronKutch** [comment]
Could I update several crates to only specify the minor version instead of patch version, when their repos indicate stricter MSRV guarantees (the last time a lot of them had their MSRV updated, it was to 1.56 because of syn 2)?

**Comment by djc** [comment]
I'm not sure what you mean? We don't want to pin dependencies for MSRV reasons -- we just keep the versions in `Cargo.lock` low enough that CI can still pass.

**Comment by AaronKutch** [comment]
In the workspace .toml, there are dependencies that specify down to the patch version, is that what pinning versions too specifically would mean?

**Comment by djc** [comment]
I think you're misunderstanding how Cargo manifest dependency specifications work. In Cargo, dependencies without an explicit operator work like they have an implicit `^` operator, so any semver-compatible newer version can be used. Since there's not currently a smallvec 2.x version, our dependency specification is already compatible with the newest version of smallvec (and if you look at `Cargo.lock` you can see that we currently use 1.13.2 for testing).

So as far as I can tell from a quick glance, the only dependency where we're actually lagging is prefix-trie, which has recently had a 0.4.0 (and 0.4.1) release which we should pick up.

**Comment by AaronKutch** [comment]
I think I have been misunderstanding how cargo updates versions this whole time. I always thought the way it worked was that if there were an array of x.y.z versions and "x" was specified, then cargo was allowed to update the y and z. If "x.y" was specified, then it could only update the z. Finally, if "x.y.z" were specified then it would fix it to that exact version. I did not realize it could jump to just any SemVer compatible version, I thought that was what the modifiers like `>=` did. It seems like `~` does what I expected, I don't know how I never noticed this. The only crates that can actually be updated right now by changing the .toml is `h3*`, `prefix-trie`, `quinn`, and `rustls`.

**Comment by djc** [comment]
Most of those are taken care of in #2217, though.

**Comment by NobodyXu** [comment]
Now that quinn, rustls and ring is upgraded, can we have a new release please?

I'm looking forward to get rid of ring 0.16 workaround for aarch64-pc-windows-msvc from cargo-binstall

Thanks for all the work on hickory-dns!

**Comment by bluejekyll** [comment]
@djc, I’m happy to punish another alpha. Anything you think we need before then?

**Comment by djc** [comment]
@bluejekyll why not just publish 0.25.0? It's been a long time.

**Comment by bluejekyll** [comment]
I can do that, though, I thought we had discussed waiting for some of the recursor work to land, we’re still waiting on the nsec3 validation for example.

**Comment by djc** [comment]
There's always more to do -- I think it'd be okay if we do another semver-incompatible release in a few months (say, after the audit is complete). For now, IMO it's worthwhile to release all the changes we've accumulated over the past 10 months and allow people to get rid of all the duplicate dependencies from the rustls stack etc.

**Comment by bluejekyll** [comment]
I'm happy to publish 0.25, but I do want to see if we can get any feedback on the new UDP -> TCP logic in the resolver. If that bakes for a week or so with no issues, then we can do 0.25, sound good?

I just published 0.25.alpha.2 today.

**Comment by NobodyXu** [comment]
Thank you!

Release 0.25 in one week, if no one complains about 0.25.alpha.2 sounds good!

I could update cargo-binstall and cut a release as well, so that people could test it!

**Comment by NobodyXu** [comment]
FYI I just published [cargo-binstall v1.10.0](https://github.com/cargo-bins/cargo-binstall/releases/tag/v1.10.0) using 0.25.0-alpha.2, if there's any bug report related to hickory-dns, I'd forward them here

**Comment by NobodyXu** [comment]
@bluejekyll cargo-binstall has been using 0.25.0-alpha and we didn't receive any issue/bug reporting regarding dns

**Comment by NobodyXu** [comment]
Would it make sense to use a more frequent/rolling release schedule?

I.e. it has been so long between release and that also make adoption of new features/bugfix longer than expected.

IMHO anything that does not change existing API probably shall not block a new 0.25.0, new features can always be added in a patch release.

**Comment by djc** [comment]
@NobodyXu what is the driver for you? Can you work with alpha/beta releases?

The reality is that no one is really funding maintenance of hickory-dns so although I've recently been spending a bunch of time during unemployment (and have been unable to find any organizations that wanted to fund my work maintaining Hickory DNS) there is always pretty limited capacity among maintainers. I think time from @bluejekyll who's effectively the lead maintainer has been especially scarce.

From the above PRs, most of these do actually change public API, only a few ones don't. From that perspective, I agree that #2437 and #2388 probably don't need to block the release, although they (along with #2428) stand in for a broader concept of robust DNSSEC.

Maybe it would be good to also have some governance discussion?

**Comment by NobodyXu** [comment]
> @NobodyXu what is the driver for you?

Mainly for cargo-binstall, drop-in replacement for cargo-install which installs pre-built binaries.

> Can you work with alpha/beta releases?

Yes, we are already using latest alpha version of hickory-dns.

> The reality is that no one is really funding maintenance of hickory-dns so although I've recently been spending a bunch of time during unemployment (and have been unable to find any organizations that wanted to fund my work maintaining Hickory DNS) 

Sorry mate, unemployment in today's economy is terrible time.

I'm willing to donate monthly to hickory-dns and send the link around social media as well.

> From the above PRs, most of these do actually change public API, only a few ones don't.

Thanks, yeah then that makes sense to postpone the 0.25 release a bit, though I do hope we can get some alpha releases.

> Maybe it would be good to also have some governance discussion?

I think we definitely would want the leadership council of rust-lang to provide some funding, or at least some help from them, like a blog post for hickory-dns to call for help and donation.

**Comment by bluejekyll** [comment]
@NobodyXu, I hear you. It's good seeing so much interest in the project, but also to leave us a little breathing room and time to discover issues before "major" releases (until we hit 1.0), my goal is to make sure that disruptive changes have time to bake. That's why we're doing alpha releases.

I'm always happy to push new alpha releases off the current mainline, but I'm uncomfortable publishing major releases as that will increase the burden on our limited contribution time.

Is the discovery of the alpha releases the primary issue that you're worried about?

**Comment by djc** [comment]
@bluejekyll can you take care of doing another alpha release? I'm not sure how you did the changelog for the last one.

**Comment by johnhtodd** [comment]
Hi - fairly new to paying attention to the project, so apologies if my once-over on the notes above is a bit naïve.

On the topic of funding (which is sadly near to my focus for $dayjob): NLNet Labs might fund a bit of this work (despite possibly a conflict of interest) because it's great to have a variety of standards-based DNS tools. Also, you may want to talk to the German Sovereign Fund, as this might be up their alley. Also perhaps Comcast, who does sponsorships for this kind of work. All these funders are very interested in encryption and security of the DNS, especially where they can be used as "libraries" for inclusion into other packages.  Do you participate at all in DNS-OARC? You may consider it, if not.

My interest in the release is also funding-related, indirectly. We (Quad9) are using Hickory-DNS embedded in Vector (https://vector.dev/) as Vector is our primary dnstap ingestion/aggregation platform. We have some funding opportunities related to DNSSEC reporting, and those in turn hinge on correct and complete EDE reporting. There was a recent PR (https://github.com/hickory-dns/hickory-dns/pull/2151) we sponsored that permits multiple EDNS messages, but I suspect the Vector team will only integrate after a formal release and not from an alpha. 

Are there any guesses as to when a "release" candidate will next be ready?  The lack of  a release gates a lot of complexity on our side for compiling custom versions of Vector.

Lastly: we (directly) might be able to put some funding together or fund some patches for development of test functionality for EDSR (https://datatracker.ietf.org/doc/draft-jt-add-dns-server-redirection/) or for authoritative DOT (server or client - https://engineering.fb.com/2018/12/21/security/dns-over-tls/ or https://blog.powerdns.com/2022/06/13/probing-dot-support-of-authoritative-servers-just-try-it) - should I start another thread on that somewhere as an issue?

**Comment by djc** [comment]
Preparing 0.25.0-alpha.3 here:

- #2561

**Comment by djc** [comment]
The 0.25.0-alpha.3 release has been published. This is really the latest and greatest version of the project, and the alpha tag merely means that we're finishing up some remaining issues and making a few more semver-incompatible changes.

**Comment by zonyitoo** [comment]
@djc When will v0.25 release? Could we expect that it will be released in early December?

**Comment by djc** [comment]
I have high hopes that it could be released in the first half of December, but no guarantees.

**Comment by zonyitoo** [comment]
Happy New Year 2025. @djc

**Comment by NobodyXu** [comment]
Feature dns-over-native-tls seems to be removed from hickory-resolver but the document isn't updated, is that intended @djc ?

**Comment by djc** [comment]
@NobodyXu not intended, would be great if you can submit a PR!

**Comment by NobodyXu** [comment]
cc @djc Based on the commits it seems that removing of it is actually intended, but the documentation didn't get updated?

https://github.com/hickory-dns/hickory-dns/commit/ca22ea794cdac221c69c6a77cbbcd38653374398
https://github.com/hickory-dns/hickory-dns/commit/df8eab62b7d82cea65b58ff1bea591b87ae57be8

To be fair, I prefer rustls over native-tls and I'm ok with sunsetting openSSL, I just want to double check if it's actually removed as it breaks my CI but the feature doc in docs.rs says it's still supported, so a bit confusing for me.

**Comment by djc** [comment]
Yes, that's what I meant -- the removal was intended, the lack of documentation changes is an oversight.

**Comment by NobodyXu** [comment]
Thanks for clarification, opened #2758

**Comment by jcgruenhage** [comment]
0.25.0 has been pushed into the future for quite a long time by now. What's actually still missing before we can have a final release candidate and then a proper release if that RC doesn't surface any major problems?

**Comment by djc** [comment]
- [ ] #2778 
- [ ] aws-lc-rs support for TLS

**Comment by djc** [comment]
I think we're pretty much ready to go with these final changes:

- [ ] #2803 
- [ ] #2805 
- [ ] #2804

**Comment by jcgruenhage** [comment]
Thanks @djc for pushing this through! :)

I see all the PRs/issues linked here merged, the version bump and changelog merged on main, but the releases aren't uploaded to crates.io yet and there is no tag yet either.

Anything we can still help with? Happy to test c141df9468b4039976ff1fcb298b6e9c84798520 in my projects for example.

**Comment by djc** [comment]
I'm basically done, @bluejekyll mentioned he wanted to have a look before release. Rest assured that I will mention here when the release has been published.

**Comment by djc** [comment]
FWIW, we're now holding the release for a fix for the #2428 regression.

**Comment by djc** [comment]
It is [done](https://github.com/hickory-dns/hickory-dns/releases/tag/v0.25.0). Sorry that it took so long -- go forth and upgrade!
